internal/tracer/exporter.go

```diff
@@ -0,0 +1,129 @@
+package tracer
+
+import (
+	"context"
+	"encoding/json"
+	"sync"
+	"time"
+
+	"github.com/windmilleng/wmclient/pkg/dirs"
+	exporttrace "go.opentelemetry.io/otel/sdk/export/trace"
+	sdktrace "go.opentelemetry.io/otel/sdk/trace"
+
+	"github.com/windmilleng/tilt/pkg/logger"
+)
+
+type Locker sync.Locker
+
+type exporter struct {
+	outgoingMu sync.Mutex
+	dir        *dirs.WindmillDir
+
+	// members for communicating with the loop() goroutine
+	spanDataCh chan *exporttrace.SpanData
+	stopCh     chan struct{}
+
+	// members only accessed by the loop() goroutine
+	queue      []*exporttrace.SpanData
+	needsFlush bool
+}
+
+// TODO(dbentley): write tests for this
+func newExporter(ctx context.Context, dir *dirs.WindmillDir) (*exporter, error) {
+	r := &exporter{
+		dir:        dir,
+		spanDataCh: make(chan *exporttrace.SpanData),
+		stopCh:     make(chan struct{}),
+	}
+	go r.loop(ctx)
+	return r, nil
+}
+
+const OutgoingFilename = "usage/outgoing.json"
+
+func (e *exporter) loop(ctx context.Context) {
+	var flushingCh chan struct{}
+	var timerCh <-chan time.Time
+	for {
+		select {
+		case sd := <-e.spanDataCh:
+			e.queue = append(e.queue, sd)
+			if timerCh == nil {
+				timerCh = time.NewTimer(5 * time.Second).C
+			}
+		case <-e.stopCh:
+			e.stopCh = nil
+			e.needsFlush = true
+		case <-timerCh:
+			timerCh = nil
+			e.needsFlush = true
+		case <-flushingCh:
+			flushingCh = nil
+		}
+
+		if !e.needsFlush || flushingCh != nil {
+			continue
+		}
+
+		if len(e.queue) == 0 {
+			e.needsFlush = false
+			continue
+		}
+
+		flushingCh = make(chan struct{})
+		go e.flush(ctx, flushingCh)
+	}
+}
+
+const maxFileSize = 32 * 1024 // 128 MiB
+
+func (e *exporter) flush(ctx context.Context, flushingCh chan struct{}) {
+	e.outgoingMu.Lock()
+	defer e.outgoingMu.Unlock()
+	defer close(flushingCh)
+	q := e.queue
```

Comment:
This read of the queue isn't thread-safe, since the queue can be simultaneously modified by your control loop

have you considered having two mutexes: one on the queue and one on the outgoing file? i think that would be simpler to reason about than your sendData channel

Reply:
I think locks can be simpler in easy cases but get more complicated quickly, vs. using a loop in a goroutine a la https://blog.golang.org/share-memory-by-communicating .

Reply:
ya, i was suggesting a possible fix to the race condition here. i don't care how you fix it.

I think you misunderstood that "Share memory by communicating" blog post. They're saying that you should pass the object to the channel, not "use channels to simulate mutually exclusive access to fields on a struct".